src/Core/Domain/Product/Image/QueryResult/Shop/ProductImageCollection.php

```diff
@@ -0,0 +1,42 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Product\Image\QueryResult\Shop;
+
+use PrestaShop\PrestaShop\Core\Data\ImmutableCollection;
+
+/**
+ * @template-extends ImmutableCollection<ProductImage>
+ */
+class ProductImageCollection extends ImmutableCollection
+{
+    public static function from(ProductImage ...$productImages)
```

Comment:
Is there a reason you need a static factory method like this? Isn't a classic constructor more simple? And why is the constructor from ImmutableCollection protected? Is it related to the template annotation and autocompletion?

Or is it maybe because we can't override the abstract class constructor and strong type it? If so maybe we actually don't need the `protected __constructor` in the abstract class? And we would only have a strong typed one in the class extending `ImmutableCollection`